shadow.3
.\" Copyright 1989, 1990, 1992, 1993, John F. Haugh II
.\" All rights reserved.
.\"
.\" Permission is granted to copy and create derivative works for any
.\" non-commercial purpose, provided this copyright notice is preserved
.\" in all copies of source code, or included in human readable form
.\" and conspicuously displayed on all copies of object code or
.\" distribution media.
.\"
.\" This software is provided on an AS-IS basis and the author makes
.\" no warrantee of any kind.
.\"
.\"	@(#)shadow.3	3.2	07:54:45	06 May 1993
.\"
.TH SHADOW 3
.SH NAME
shadow \- encrypted password file routines
.SH SYNTAX
#include <shadow.h>
.br
struct spwd *getspent();
.br
struct spwd *getspnam(char * name);
.br
void setspent();
.br
void endspent();
.br
struct spwd *fgetspent(FILE *fp);
.br
struct spwd *sgetspent(char *cp);
.br
int putspent(struct spwd *p,FILE *fp);
.br
int lckpwdf ();
.br
int ulckpwdf ();
.SH DESCRIPTION
.I shadow
manipulates the contents of the shadow password file,
\fB/etc/shadow\fR.
The structure in the \fI#include\fR file is
.sp
struct spwd {
.in +.5i
.br
	char	*sp_namp; /* user login name */
.br
	char	*sp_pwdp; /* encrypted password */
.br
	long	sp_lstchg; /* last password change */
.br
	int	sp_min; /* days until change allowed. */
.br
	int	sp_max; /* days before change required */
.br
	int	sp_warn; /* days warning for expiration */
.br
	int	sp_inact; /* days before account inactive */
.br
	int	sp_expire; /* date when account expires */
.br
	int	sp_flag; /* reserved for future use */
.br
.in -.5i
}
.PP
The meanings of each field are
.sp
sp_namp \- pointer to null-terminated user name.
.br
sp_pwdp \- pointer to null-terminated password.
.br
sp_lstchg \- days since Jan 1, 1970 password was last changed.
.br
sp_min \- days before which password may not be changed.
.br
sp_max \- days after which password must be changed.
.br
sp_warn \- days before password is to expire that user is warned
of pending password expiration.
.br
sp_inact \- days after password expires that account is considered
inactive and disabled.
.br
sp_expire \- days since Jan 1, 1970 when account will be disabled.
.br
sp_flag \- reserved for future use.
.SH DESCRIPTION
\fIgetspent\fR, \fIgetspname\fR, \fIfgetspent\fR, and \fIsgetspent\fR
each return a pointer to a \fBstruct spwd\fR.
\fIgetspent\fR returns the
next entry from the file, and \fIfgetspent\fR returns the next
entry from the given stream, which is assumed to be a file of
the proper format.
\fIsgetspent\fR returns a pointer to a \fBstruct spwd\fR using the
provided string as input.
\fIgetspnam\fR searches from the current position in the file for
an entry matching \fIname\fR.
.PP
\fIsetspent\fR and \fIendspent\fR may be used to begin and end,
respectively, access to the shadow password file.
.PP
The \fIlckpwdf\fR and \fIulckpwdf\fR routines should be used to
insure exclusive access to the \fB/etc/shadow\fR file.
\fIlckpwdf\fR attempts to acquire a lock using \fIpw_lock\fR for
up to 15 seconds.
It continues by attempting to acquire a second lock using \fIspw_lock\fR
for the remainder of the initial 15 seconds.
Should either attempt fail after a total of 15 seconds, \fIlckpwdf\fR
returns -1.
When both locks are acquired 0 is returned.
.SH DIAGNOSTICS
Routines return NULL if no more entries are available or if an
error occurs during processing.
Routines which have \fBint\fR as the return value return 0 for
success and -1 for failure.
.SH CAVEATS
These routines may only be used by the super user as access to
the shadow password file is restricted.
.SH FILES
/etc/shadow \- encrypted user passwords
.SH SEE ALSO
getpwent(3),
shadow(4)


NEWS
$Id$

shadow-4.1.5.1 -> shadow-4.2					UNRELEASED

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after
    sending a SIGTERM), and kill it only if it did not terminate by itself.
    No delay will be enforced if the child cooperates.
  * Default ENV_SUPATH is /sbin:/bin:/usr/sbin:/usr/bin
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.

shadow.c
/*
 * Copyright 1989, 1990, 1991, 1992, John F. Haugh II
 * All rights reserved.
 *
 * Permission is granted to copy and create derivative works for any
 * non-commercial purpose, provided this copyright notice is preserved
 * in all copies of source code, or included in human readable form
 * and conspicuously displayed on all copies of object code or
 * distribution media.
 */

#include "shadow.h"
#include "config.h"
#include <stdio.h>

#ifdef	STDLIB_H
#include <stdlib.h>
#endif

#ifndef	BSD
#include <string.h>
#include <memory.h>
#else
#include <strings.h>
#define	strchr	index
#define	strrchr	rindex
#endif

#ifdef	NDBM
#include <ndbm.h>
#include <fcntl.h>
DBM	*sp_dbm;
int	sp_dbm_mode = -1;
static	int	dbmopened;
static	int	dbmerror;
#endif

#ifndef	lint
static	char	sccsid[] = "@(#)shadow.c	3.13	08:07:15	19 Jul 1993";
#endif

static	FILE	*shadow;
static	char	spwbuf[BUFSIZ];
static	struct	spwd	spwd;

#define	FIELDS	9
#define	OFIELDS	5

void
setspent ()
{
	if (shadow)
		rewind (shadow);
	else
		shadow = fopen (SHADOW, "r");

	/*
	 * Attempt to open the DBM files if they have never been opened
	 * and an error has never been returned.
	 */

#ifdef NDBM
	if (! dbmerror && ! dbmopened) {
		int	mode;
		char	dbmfiles[BUFSIZ];

		strcpy (dbmfiles, SHADOW);
		strcat (dbmfiles, ".pag");

		if (sp_dbm_mode == -1)
			mode = O_RDWR;
		else
			mode = (sp_dbm_mode == O_RDWR) ? O_RDWR:O_RDONLY;

		if (! (sp_dbm = dbm_open (SHADOW, mode, 0)))
			dbmerror = 1;
		else
			dbmopened = 1;
	}
#endif
}

void
endspent ()
{
	if (shadow)
		(void) fclose (shadow);

	shadow = (FILE *) 0;
#ifdef	NDBM
	if (dbmopened && sp_dbm) {
		dbm_close (sp_dbm);
		sp_dbm = 0;
	}
	dbmopened = 0;
	dbmerror = 0;
#endif
}

struct spwd *
sgetspent (string)
char	*string;
{
	char	*fields[FIELDS];
	char	*cp;
	char	*cpp;
	int	atoi ();
	long	atol ();
	int	i;

	strncpy (spwbuf, string, BUFSIZ-1);
	spwbuf[BUFSIZ-1] = '\0';

	if (cp = strrchr (spwbuf, '\n'))
		*cp = '\0';

	for (cp = spwbuf, i = 0;*cp && i < FIELDS;i++) {
		fields[i] = cp;
		while (*cp && *cp != ':')
			cp++;

		if (*cp)
			*cp++ = '\0';
	}
	if (i == (FIELDS-1))
		fields[i++] = cp;

	if ((cp && *cp) || (i != FIELDS && i != OFIELDS))
		return 0;

	spwd.sp_namp = fields[0];
	spwd.sp_pwdp = fields[1];

	if ((spwd.sp_lstchg = strtol (fields[2], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[2][0] == '\0')
		spwd.sp_lstchg = -1;

	if ((spwd.sp_min = strtol (fields[3], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[3][0] == '\0')
		spwd.sp_min = -1;

	if ((spwd.sp_max = strtol (fields[4], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[4][0] == '\0')
		spwd.sp_max = -1;

	if (i == OFIELDS) {
		spwd.sp_warn = spwd.sp_inact = spwd.sp_expire =
			spwd.sp_flag = -1;

		return &spwd;
	}
	if ((spwd.sp_warn = strtol (fields[5], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[5][0] == '\0')
		spwd.sp_warn = -1;

	if ((spwd.sp_inact = strtol (fields[6], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[6][0] == '\0')
		spwd.sp_inact = -1;

	if ((spwd.sp_expire = strtol (fields[7], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[7][0] == '\0')
		spwd.sp_expire = -1;

	if ((spwd.sp_flag = strtol (fields[8], &cpp, 10)) == 0 && *cpp)
		return 0;
	else if (fields[8][0] == '\0')
		spwd.sp_flag = -1;

	return (&spwd);
}

struct spwd
*fgetspent (fp)
FILE	*fp;
{
	char	buf[BUFSIZ];

	if (! fp)
		return (0);

	if (fgets (buf, BUFSIZ, fp) == (char *) 0)
		return (0);

	return sgetspent (buf);
}

struct spwd
*getspent ()
{
	if (! shadow)
		setspent ();

	return (fgetspent (shadow));
}

struct spwd
*getspnam (name)
#if	__STDC__
const
#endif
char	*name;
{
	struct	spwd	*sp;
#ifdef NDBM
	datum	key;
	datum	content;
#endif

	setspent ();

#ifdef NDBM

	/*
	 * If the DBM file are now open, create a key for this UID and
	 * try to fetch the entry from the database.  A matching record
	 * will be unpacked into a static structure and returned to
	 * the user.
	 */

	if (dbmopened) {
		key.dsize = strlen (name);
		key.dptr = name;

		content = dbm_fetch (sp_dbm, key);
		if (content.dptr != 0) {
			memcpy (spwbuf, content.dptr, content.dsize);
			spw_unpack (spwbuf, content.dsize, &spwd);
			return &spwd;
		}
	}
#endif
	while ((sp = getspent ()) != (struct spwd *) 0) {
		if (strcmp (name, sp->sp_namp) == 0)
			return (sp);
	}
	return (0);
}

int
putspent (sp, fp)
#if	__STDC__
const
#endif
struct	spwd	*sp;
FILE	*fp;
{
	int	errors = 0;

	if (! fp || ! sp)
		return -1;

	if (fprintf (fp, "%s:%s:", sp->sp_namp, sp->sp_pwdp) < 0)
		errors++;

	if (sp->sp_lstchg != -1) {
		if (fprintf (fp, "%ld:", sp->sp_lstchg) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_min != -1) {
		if (fprintf (fp, "%ld:", sp->sp_min) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_max != -1) {
		if (fprintf (fp, "%ld:", sp->sp_max) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_warn != -1) {
		if (fprintf (fp, "%ld:", sp->sp_warn) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_inact != -1) {
		if (fprintf (fp, "%ld:", sp->sp_inact) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_expire != -1) {
		if (fprintf (fp, "%ld:", sp->sp_expire) < 0)
			errors++;
	} else if (putc (':', fp) == EOF)
		errors++;

	if (sp->sp_flag != -1) {
		if (fprintf (fp, "%ld", sp->sp_flag) < 0)
			errors++;
	}
	if (putc ('\n', fp) == EOF)
		errors++;

	if (errors)
		return -1;
	else
		return 0;
}
